host install libvirt

SeanMooney opened the pull request:
refactor libvirt to be host installed

Closes: [OSPRH-960](https://issues.redhat.com//browse/OSPRH-960)
Depends-On: https://github.com/openstack-k8s-operators/ci-framework/pull/910

fultonj commented on roles/edpm_libvirt/defaults/main.yml:
```diff
@@ -26,10 +26,27 @@ edpm_libvirt_image_download_delay: 5
+  # for qemu
+  - ceph-common
+
 edpm_libvirt_ceph_path: /var/lib/openstack/config/ceph/
```
Hopefully libvirt can use this path. If it can't for some reason we can work something out to keep ceph client and ceph server separate, e.g. ceph server can use `/etc/ceph/$FSID`.

gibizer commented:
the libvirt molecule failure seems relevant
```
2023-10-19 15:44:54.508615 | controller | TASK [osp.edpm.edpm_libvirt : Install libvirt packages] ************************
2023-10-19 15:45:57.524336 | controller | failed: [compute-1] (item=ceph-common) => changed=false
2023-10-19 15:45:57.524405 | controller |   ansible_loop_var: item
2023-10-19 15:45:57.524427 | controller |   failures:
2023-10-19 15:45:57.526001 | controller |   - No package ceph-common available.
2023-10-19 15:45:57.526039 | controller |   item: ceph-common
2023-10-19 15:45:57.526050 | controller |   msg: Failed to install some of the specified packages
2023-10-19 15:45:57.526059 | controller |   rc: 1
```

gibizer commented:
The deploy job failure also seems relevant:
```
TASK [Download images for edpm_libvirt role] ***********************************
Thursday 19 October 2023  21:02:30 +0000 (0:00:00.709)       0:00:33.044 ****** 
[0;31mERROR! Could not find specified file in role: tasks/download_cache.yml[0m
```
https://logserver.rdoproject.org/67/467/0fd390bbf14be0ea83788688ab3d75656c71fa9f/github-check/podified-multinode-edpm-deployment-crc/30d8404/controller/ci-framework-data/logs/quay-io-openstack-k8s-operators-openstack-must-gather-sha256-923bc944999f722935616f4e3fa620f5a1b037c8cef824258f2bad11f0a178e4/namespaces/openstack/pods/dataplane-deployment-download-cache-openstack-edpm-2mm42/logs/openstackansibleee.log

gibizer commented on roles/edpm_libvirt/molecule/default/prepare.yml:
```diff
@@ -15,15 +15,28 @@
 - name: Setup DUT
   hosts: all
   tasks:
+    - name: Enable package repos
```
I guess we need this in the cache role to allow caching the package

SeanMooney replied:
so we only need this for ceph-common
its not available by default it seams on rocky/centos

alghouth in general on rhel its what would configure the downstream repo for our packages in general

gibizer commented on roles/edpm_libvirt/tasks/download_cache.yml:
We probably need a related change in the download_cache role as it still tries to use this file

SeanMooney replied:
ill readd it and just make it pull the rpms

gibizer reviewed (commented):
I like the direction here we just make it work :)

SeanMooney commented:
> > the libvirt molecule failure seems relevant
> > ```
> > 2023-10-19 15:44:54.508615 | controller | TASK [osp.edpm.edpm_libvirt : Install libvirt packages] ************************
> > 2023-10-19 15:45:57.524336 | controller | failed: [compute-1] (item=ceph-common) => changed=false
> > 2023-10-19 15:45:57.524405 | controller |   ansible_loop_var: item
> > 2023-10-19 15:45:57.524427 | controller |   failures:
> > 2023-10-19 15:45:57.526001 | controller |   - No package ceph-common available.
> > 2023-10-19 15:45:57.526039 | controller |   item: ceph-common
> > 2023-10-19 15:45:57.526050 | controller |   msg: Failed to install some of the specified packages
> > 2023-10-19 15:45:57.526059 | controller |   rc: 1
> > ```
> 
> Repo centos9-storage provides ceph-common. The commands in this script should set up said repo:
> 
> https://github.com/openstack-k8s-operators/install_yamls/blob/93e9be4e0fb1bb38b20b34cc6a86fc2889cfa3fa/devsetup/scripts/edpm-compute-repos.sh
> 
> One solution is for the DUT to use the same commands to prepare its repo since that's what upstream EDPM nodes will use.
> 
> The above script was removed in favor of the edpm-service `repo-setup` which basically runs the same commands within ansible.

i coppied the repo setup command that i added to prepare from install yamls

https://github.com/openstack-k8s-operators/install_yamls/blob/main/devsetup/edpm/services/dataplane_v1beta1_openstackdataplaneservice_reposetup.yaml#L15C2-L29C35

that is the one that runs if you use install yamls to deploy an edpm host

and in the molecule env it ran to completion 
https://review.rdoproject.org/zuul/build/406fdd4d95de412695f6bb56f023fdbe/log/job-output.txt#381-491

its just the verify tests which i have not updated that actually failed.

i added the repo setup to prepare as that is something that is meant to be done before the libvirt role is used. its actually meant to be done before any other dataplane service runs so the cache role should not be using it.

gibizer commented:
It seems the vagrant provider has some issues.
```
2023-10-20 13:52:39.219258 | controller | INFO     Verifier completed successfully.
2023-10-20 13:52:39.220485 | controller | INFO     Performing prerun with role_name_check=0...
2023-10-20 13:52:39.222305 | controller | INFO     Set ANSIBLE_LIBRARY=/home/zuul/.cache/ansible-compat/40c08c/modules:/home/zuul/.ansible/plugins/modules:/usr/share/ansible/plugins/modules:/home/zuul/.cache/ansible-compat/40c08c/modules:/home/zuul/.ansible/plugins/modules:/usr/share/ansible/plugins/modules
2023-10-20 13:52:39.223761 | controller | INFO     Set ANSIBLE_COLLECTIONS_PATH=/home/zuul/.cache/ansible-compat/40c08c/collections:/home/zuul/.cache/ansible-compat/40c08c/collections:/home/zuul/.ansible/collections:/usr/share/ansible/collections:/home/zuul/.cache/ansible-compat/40c08c/collections:/home/zuul/.ansible/collections:/usr/share/ansible/collections
2023-10-20 13:52:39.225098 | controller | INFO     Set ANSIBLE_ROLES_PATH=/home/zuul/.cache/ansible-compat/40c08c/roles:/home/zuul/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles:/home/zuul/.cache/ansible-compat/40c08c/roles:/home/zuul/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles
2023-10-20 13:52:39.226995 | controller | INFO     Running from /home/zuul/src/github.com/openstack-k8s-operators/edpm-ansible/roles/edpm_libvirt : ansible-galaxy collection install -vvv --force ../..
2023-10-20 13:52:40.530503 | controller | INFO     Running vagrant > prepare
2023-10-20 13:52:40.532144 | controller | CRITICAL vagrant executable was not found!
2023-10-20 13:52:40.532614 | controller | WARNING  An error occurred during the test sequence action: 'prepare'. Cleaning up.
```

gibizer commented:
And in the deploy job:
```

TASK [osp.edpm.edpm_container_manage : Create containers managed by Podman for /var/lib/openstack/config/containers] ***
Friday 20 October 2023  19:09:13 +0000 (0:00:00.085)       0:00:23.510 ******** 
[1;35m[WARNING]: ERROR: Can't run container nova_compute stderr:[0m
[1;35mtime="2023-10-20T19:09:14Z" level=info msg="podman filtering at log level info"[0m
[1;35mtime="2023-10-20T19:09:14Z" level=info msg="Not using native diff for overlay,[0m
[1;35mthis may cause degraded performance for building images: kernel has[0m
[1;35mCONFIG_OVERLAY_FS_REDIRECT_DIR enabled" time="2023-10-20T19:09:14Z" level=info[0m
[1;35mmsg="Setting parallel job count to 25" time="2023-10-20T19:09:14Z" level=info[0m
[1;35mmsg="Sysctl net.ipv4.ping_group_range=0 0 ignored in containers.conf, since[0m
[1;35mNetwork Namespace set to host" Error: statfs /var/lib/openstack/config/ceph: no[0m
[1;35msuch file or directory[0m
[0;31mfatal: [edpm-compute-0]: FAILED! => {"changed": false, "msg": "Failed containers: nova_compute"}[0m
```
So as we discussed we need to make sure the the ceph information is still available to nova even if libvirt is not containerized.

bogdando commented:
I'd move ffu from the blocking path and merge that we have

bogdando commented:
> the libvirt molecule failure seems relevant
> 
> ```
> 2023-10-19 15:44:54.508615 | controller | TASK [osp.edpm.edpm_libvirt : Install libvirt packages] ************************
> 2023-10-19 15:45:57.524336 | controller | failed: [compute-1] (item=ceph-common) => changed=false
> 2023-10-19 15:45:57.524405 | controller |   ansible_loop_var: item
> 2023-10-19 15:45:57.524427 | controller |   failures:
> 2023-10-19 15:45:57.526001 | controller |   - No package ceph-common available.
> 2023-10-19 15:45:57.526039 | controller |   item: ceph-common
> 2023-10-19 15:45:57.526050 | controller |   msg: Failed to install some of the specified packages
> 2023-10-19 15:45:57.526059 | controller |   rc: 1
> ```

if we have to install ceph-common on the host and elsewhere in ceph-specific containers, wouldn't this bring ABI mismatching problems?

bogdando commented on roles/edpm_libvirt/tasks/configure.yml:
```diff
@@ -13,54 +13,46 @@
     mode: "{{ item.mode | default(omit) }}"
   loop:
-    - {"path": "/var/lib/openstack/config/libvirt"}
+    - {"path": "/etc/libvirt", "owner": "root", "group": "root", "mode": "0755"}
```
why do we need to change the rpm shipped permissions?

SeanMooney replied:
we are creating it before the rpm is installed
but also the rpm does not create a libvirt user it runs libvirt as root
so this is the same permission it would have anyway

bogdando replied:
Let's ensure RPM installed before doing that?
Let's run it as root? My concern is libvirt team could tell us later they don't test running it as non-root, like they don't for running it w/o systemd containerd integrated cgroups management. And we will on our own support again

SeanMooney replied:
ya we talked about this else where.
im going to reoder the install to happne before configure and defer starting the servics until after.
that should also resolve the selinux isues as the folder will be created with the correct se linux lables.

bogdando commented on roles/edpm_libvirt/molecule/default/verify.yml:
```diff
@@ -13,38 +13,46 @@
-        - "libvirt_virtproxyd"
-        - "libvirt_virtqemud"
-        - "libvirt_virtsecretd"
+        # rhel installed services have service units defiend via rpm in
```
could we use the service units shipped by the RHEL package and remove our customizations?

SeanMooney replied:
we are i have removed our custom services and we are using the ones provide by the rhel rpm which are not in /etc/systemd which is why i left this comment

SeanMooney replied:
to be clear the only time we are not using the libvirt packaged service files for libvirt is libvirt_guest and that is because we only use the shutdown part of libvirt-guest not the start up part so we have to override it

SeanMooney commented on roles/edpm_libvirt/tasks/configure.yml:
```diff
@@ -13,54 +13,46 @@
+- name: Disable selinux for now
   tags:
-    - configure
+    - install
```
oh this should be configure thats a copy paste bug

SeanMooney commented:
this is jsut a rebase ill be reworking this today

SeanMooney commented:
/hold gibi is going to try and build this and test it end to end via the dataplane operator

SeanMooney commented on roles/edpm_libvirt/molecule/vagrant/molecule.yml:
the ci framework currently runs all scenarios  and this one will not work so I'm going
to pull this out into its own pr.

gibizer commented:
I tested this with a local build `quay.io/gibi/openstack-ansibleee-runner:v0.0.50` and saw two issues:
* nova_compute container fails to start with `Error: statfs /var/lib/openstack/config/ceph: no
such file or directory` ( I fixed it manually)
* VM cannot be started as libvirt fails with `libvirt.libvirtError: Cannot create daemon common directory '/run/libvirt/common': Not a directory` but there is directory:
  ```
  [root@edpm-compute-0 ~]# ll /run/libvirt/common
   total 4
   -rw-------. 1 root root 32 Nov 13 17:28 system.token
```

SeanMooney commented:
> > the libvirt molecule failure seems relevant
> > ```
> > 2023-10-19 15:44:54.508615 | controller | TASK [osp.edpm.edpm_libvirt : Install libvirt packages] ************************
> > 2023-10-19 15:45:57.524336 | controller | failed: [compute-1] (item=ceph-common) => changed=false
> > 2023-10-19 15:45:57.524405 | controller |   ansible_loop_var: item
> > 2023-10-19 15:45:57.524427 | controller |   failures:
> > 2023-10-19 15:45:57.526001 | controller |   - No package ceph-common available.
> > 2023-10-19 15:45:57.526039 | controller |   item: ceph-common
> > 2023-10-19 15:45:57.526050 | controller |   msg: Failed to install some of the specified packages
> > 2023-10-19 15:45:57.526059 | controller |   rc: 1
> > ```
> 
> if we have to install ceph-common on the host and elsewhere in ceph-specific containers, wouldn't this bring ABI mismatching problems?

no, we are already installing ceph-common on the host and in the contianer anyway.
its install to provide the udev rules and secondary to provided ceph/rbd client cli for debuging.

this is not being used from the host in the container, the container uses librbd and the ceph client it has but does not communicate with cpeh via a binary interface.

we have been host installing ceph-common since we first contaierised OpenStack without any abi issues to this point.

bogdando commented on roles/edpm_libvirt/defaults/main.yml:
```diff
@@ -20,16 +20,34 @@
+  - edk2-ovmf
+  # ceph-common is needed for ceph support to provide the ceph client lib
+  # for qemu and the ceph udev rules
+  - ceph-common
```
please double-check if that is aligned with that we have today in TCIB:
```
tcib_packages:
  common:
  - ceph-common                
  - cyrus-sasl-scram (?)
  - jq (not needed)
  - libguestfs +
  - libseccomp +
  - libvirt-admin +
  - libvirt-client +
  - libvirt-daemon +
  - libvirt-daemon-config-nwfilter (?)
  - libvirt-daemon-driver-nwfilter (?)
  - openssl-perl (unrelated, but needs to be present on host)
  - openstack-nova-migration (only if we dropping it for osp18)
  - openvswitch (unrelated)
  - podman (not needed)
  - qemu-kvm +
  - swtpm +
  - swtpm-tools +
  x86_64:
  - edk2-ovmf +
```

bogdando commented on roles/edpm_libvirt/meta/argument_specs.yml:
```diff
@@ -11,16 +11,28 @@ argument_specs:
+          - swtpm
+          - swtpm-tools
+          - edk2-ovmf
+          - ceph-common
```
ditto

bogdando commented on roles/edpm_libvirt/tasks/configure.yml:
```diff
@@ -13,54 +13,65 @@
+  ansible.builtin.template:
+    src: "{{ item.src }}"
+    dest: "/etc/libvirt/{{ item.dest }}"
+    mode: "0644"
```
it shouldn't be world read

SeanMooney replied:
[root@edpm-compute-0 ~]# ls -alZ /etc/libvirt/
total 144
drwx------.   6 root root system_u:object_r:virt_etc_t:s0      4096 Nov 21 12:35 .
drwxr-xr-x. 114 root root system_u:object_r:etc_t:s0           8192 Nov 27 12:34 ..
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0       450 Jul  3 10:08 libvirt-admin.conf
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0       547 Jul  3 10:08 libvirt.conf
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0     17826 Aug 25 08:54 libvirtd.conf
drwx------.   2 root root system_u:object_r:virt_etc_rw_t:s0   4096 Nov 16 14:32 nwfilter
drwx------.   4 root root system_u:object_r:virt_etc_rw_t:s0     39 Nov 23 09:10 qemu
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0       242 Nov 16 14:33 qemu.conf
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0      2169 Aug 25 08:54 qemu-lockd.conf
drwx------.   2 root root system_u:object_r:virt_etc_rw_t:s0      6 Aug 25 08:55 secrets
drwx------.   3 root root system_u:object_r:virt_etc_rw_t:s0     23 Nov 16 14:32 storage
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0     12421 Aug 25 08:54 virtinterfaced.conf
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0      3058 Jul  3 10:08 virtlockd.conf
-rw-r--r--.   1 root root system_u:object_r:virtlogd_etc_t:s0   106 Nov 16 14:33 virtlogd.conf
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0     12411 Aug 25 08:54 virtnetworkd.conf
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0       235 Nov 16 14:33 virtnodedevd.conf
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0     12416 Aug 25 08:54 virtnwfilterd.conf
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0       259 Nov 16 14:33 virtproxyd.conf
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0       232 Nov 21 12:35 virtqemud.conf
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0       234 Nov 16 14:33 virtsecretd.conf
-rw-r--r--.   1 root root system_u:object_r:virt_etc_t:s0     12411 Aug 25 08:54 virtstoraged.conf


even the files that come form the rpm are 644 by default just an fyi
i do not modify qemu-lockd.conf or libvirtd.conf  above so they have the rpm defaualt content and mode.

i can drop the world read but 0644 is the expected mode for files in /etc/ in general

ill update it to make this 0640 but that will be a change in behavior form how this is normally done.

bogdando reviewed (changes requested):
please fix conf files permisions

gibizer commented:
I deployed this, built and installed the selinux fix and stopped, disabled, and masked virtlockd.socket. Then I tried to live migrate and got the same error:
```
2023-11-30 15:03:22.526 2 ERROR nova.compute.manager [instance: 524f82f0-cb66-4e88-ab3b-1cd2179b7765] libvirt.libvirtError: Timed out during operation: cannot acquire state change lock (held by monitor=remoteDispatchDomainMigratePerform3Params)
```

gibizer commented:
According to my tests decontainerized libvirt works. I can boot VM, cold migrate VM. For live-migration to work we need an extra firewall rule I proposed in https://github.com/openstack-k8s-operators/edpm-ansible/pull/513

gibizer reviewed (commented):
This looks good to me

bogdando commented on roles/edpm_libvirt/handlers/main.yml:
```diff
@@ -14,15 +14,15 @@
-  loop: "{{ edpm_libvirt_containers }}"
+    daemon_reload: true
+  loop: "{{ edpm_libvirt_services }}"
   listen: "Restart libvirt"
```
is there a special need to listen for the different value than the name has?

bogdando commented on roles/edpm_libvirt/templates/edpm_libvirt_guests.service:
```diff
@@ -10,8 +10,8 @@ Documentation=https://libvirt.org
 EnvironmentFile=-/etc/sysconfig/libvirt-guests
-ExecStart=/bin/podman exec libvirt_virtproxyd /bin/rm -f /var/lib/libvirt/libvirt-guests
-ExecStop=/bin/podman exec libvirt_virtproxyd /bin/sh -x /usr/libexec/libvirt-guests.sh shutdown
+ExecStart=/bin/rm -f /var/lib/libvirt/libvirt-guests
```
could we use the default unit shipped with the libvirt-daemon/-common packages? We had to maintain out copy because of the need to wrap it with podman. But we no longer need to do this after de-containerization.

SeanMooney replied:
we also need the addtion fo Before=edpm_nova_compute.service

i could use dropin files to get a similar effect but i fell like that is more likely to break going forward.

bogdando commented on roles/edpm_libvirt/templates/edpm_libvirt.target:
```diff
@@ -1,11 +1,11 @@
-After=edpm_libvirt_virtproxyd.service
-After=edpm_libvirt_virtqemud.service
-After=edpm_libvirt_virtsecretd.service
+Wants=virtlogd.service
```
ditto, let's use the target defined by the package.
We don't want to maintain a copy of it here

bogdando commented on roles/edpm_libvirt/tasks/post-install.yml:
```diff
@@ -57,6 +73,23 @@
+        dest: "/etc/ceph/{{ (item.path | basename) }}"
+        owner: root
+        group: root
+        mode: "0644"
```
please don't set this world read. @fultonj WDYT?

SeanMooney replied:
again world reabale is what rhel would do by default
we can make it stricter but when we do we risk breaking things by deviating form the norm.

by default file in /etc are often 0644

but yes i can update his if @fultonj prefers a different value.

bogdando reviewed (changes requested):
please see inline

bogdando commented:
lgtm

bogdando reviewed (commented):
lgtm

gibizer commented:
/hold we dont have https://github.com/redhat-openstack/openstack-selinux/pull/106 available to us as a package

slagle commented on roles/edpm_libvirt/tasks/download_cache.yml:
this should be included in roles/edpm_download_cache/tasks/packages.yml

slagle replied:
and removed from roles/edpm_download_cache/tasks/container_images.yml

SeanMooney commented:
/unhold

this has now promoted upstream and downstream
openstack-selinux-0.8.37-0.20231201082840.165e251.el9

bogdando commented:
/LGTM

SeanMooney commented:
check-rdo

SeanMooney commented:
recheck

check-rdo shoudl work based on https://github.com/rdo-infra/review.rdoproject.org-config/blob/master/zuul.d/github.yaml#L28C37-L28C46

but this is not triggering. i think its current state is because zuul was restart while the job was running in the past.

SeanMooney commented:
check-rdo

gibizer commented:
> /unhold
> 
> this has now promoted upstream and downstream openstack-selinux-0.8.37-0.20231201082840.165e251.el9

I confirm. A runner built from this PR now can live-migrate without any manual selinux WA
